core/src/main/java/hudson/tasks/Fingerprinter.java

```diff
@@ -92,13 +93,40 @@
      */
     private final String targets;
 
+    /**
+     * Default null 'excludes' pattern as in Ant.
+     */
+    private String excludes = null;
+
+    /**
+     * Default ant exclusion
+     */
+    private Boolean defaultExcludes = true;
+
+    /**
+     * Indicate whether include and exclude patterns should be considered as case sensitive
+     */
+    private Boolean caseSensitive = true;
+
     @Deprecated
     Boolean recordBuildArtifacts;
 
     @DataBoundConstructor public Fingerprinter(String targets) {
         this.targets = targets;
     }
 
+    @DataBoundSetter public void setExcludes(String excludes) {
+        this.excludes = excludes;
+    }
+
+    @DataBoundSetter public void setDefaultExcludes(Boolean defaultExcludes) {
```

Comment:
Getters and setters should be of type `boolean` not `Boolean`.

Please reread https://github.com/jenkinsci/jenkins/pull/3915#discussion_r262178204. This is, if anything, worse since now after deserializing an old `config.xml` your code

```java
new FindRecords(targets, excludes, defaultExcludes, caseSensitive, build.getTimeInMillis())
```

will throw a `NullPointerException` unboxing e.g. a null `caseSensitive`.

Reply:
By the way [this](https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#handling-default-values) may be helpful.

Reply:
@jglick You have mentioned that we can use `readResolve` method to mitigate issue of passing null Boolean object.
- I found a readResolve method in ArtiactArchiver.java file. (Path: hudson/tasks/ArtifactArchiver.java)
```
// Backwards compatibility for older builds
    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
            justification = "Null checks in readResolve are valid since we deserialize and upgrade objects")
    protected Object readResolve() {
        if (allowEmptyArchive == null) {
            this.allowEmptyArchive = SystemProperties.getBoolean(ArtifactArchiver.class.getName()+".warnOnEmpty");
        }
        if (defaultExcludes == null){
            defaultExcludes = true;
        }
        if (caseSensitive == null) {
            caseSensitive = true;
        }
        return this;
    }
```
Based on it, I guess that following code can be written. But I don't understand how and when readResolve method gets called.
```
// Backwards compatibility for older builds
    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
            justification = "Null checks in readResolve are valid since we deserialize and upgrade objects")
    protected Object readResolve() {
        if (defaultExcludes == null){
            defaultExcludes = true;
        }
        if (caseSensitive == null) {
            caseSensitive = true;
        }
        return this;
    }
```

Reply:
Yes, you can `readResolve` if that is the most convenient thing. It is called during XStream deserialization, in this case of a `config.xml` for a freestyle-ish project which has been configured via the GUI in a previous Jenkins session. (Pipeline projects would be creating the object via its `@DataBoundConstructor` and `@DataBoundSetter`.) It is best to read the Java serialization guide to understand what methods like this do.

A somewhat less magical alternative is to leave the field `null` and then interpret a default value in the getter and make sure all other accesses go through the getter. You still need to understand serialization: for a freestyle project, the constructor will be called only when the project is (re-)configured, not every time Jenkins starts, and this means for example that an instance initializer

```java
private Boolean caseSensitive = true;
```

will never have been called for a publisher which had been configured prior to the Jenkins upgrade represented by this patch.

If in doubt, you can always write functional tests using `@LocalData` which reproduce this kind of scenario.

Reply:
@jglick I added readResolve method below. I have tagged you there. You can also check my last commit-7.